src/methods.jl

```diff
@@ -10,318 +19,446 @@ derivatives.
+    return (f, x) -> condition * maximum(abs, f(x))
+end
 
-# Estimate the bound on the function value and its derivatives at a point
-_estimate_bound(x, cond) = add_tiny(cond * maximum(abs, x))
+function Base.show(io::IO, m::MIME"text/plain", x::FiniteDifferenceMethod)
+    @printf io "FiniteDifferenceMethod:\n"
+    @printf io "  order of method:       %d\n" length(x.grid)
+    @printf io "  order of derivative:   %d\n" x.q
+    @printf io "  grid:                  %s\n" x.grid
+    @printf io "  coefficients:          %s\n" x.coefs
+end
 
 """
-    fdm(m::FiniteDifferenceMethod, f, x[, Val(false)]; kwargs...) -> Real
-    fdm(m::FiniteDifferenceMethod, f, x, Val(true); kwargs...) -> Tuple{FiniteDifferenceMethod, Real}
-
-Compute the derivative of `f` at `x` using the finite differencing method `m`.
-The optional `Val` argument dictates whether the method should be returned alongside the
-derivative value, which can be useful for examining the step size used and other such
-parameters.
-
-The recognized keywords are:
+    function estimate_step(
+        m::FiniteDifferenceMethod,
+        f::Function,
+        x::T;
+        factor::Int=1,
+        max_step=convert(T, 0.1)
+    ) where T<:AbstractFloat
+
+Estimate the step size for a finite difference method `m`. Also estimates the error of the
+estimate of the derivative.
+
+# Arguments
+- `m::FiniteDifferenceMethod`: Finite difference method to estimate the step size for.
+- `f::Function`: Function to evaluate the derivative of.
+- `x::T`: Point to estimate the derivative at.
+
+# Keywords
+- `factor::Int=1`. Factor to amplify the estimated round-off error by. This can be used
+    to force a more conservative step size.
+- `max_step=convert(T, 0.1)`: Maximum step size.
+
+# Returns
+- `Tuple{T, T}`: Estimated step size and an estimate of the error of the finite difference
+    estimate.
+"""
+function estimate_step(
+    m::FiniteDifferenceMethod,
+    f::Function,
+    x::T;
+    factor::Int=1,
+    max_step=convert(T, 0.1)
+) where T<:AbstractFloat
+    p = length(m.coefs)
+    q = m.q
+    f_x = float(f(x))
 
-* `adapt`: The number of adaptive steps to use improve the estimate of `bound`.
-* `bound`: Bound on the value of the function and its derivatives at `x`.
-* `condition`: The condition number. See [`DEFAULT_CONDITION`](@ref).
-* `eps`: The assumed roundoff error. Defaults to `add_tiny(eps())`.
+    # Estimate the bound and round-off error.
+    ε = add_tiny(maximum(eps, f_x)) * factor
+    M = add_tiny(m.bound_estimator(f, x))
 
-!!! warning
-    Bounds can't be adaptively computed over nonstandard grids; passing a value for
-    `adapt` greater than 0 when `m::Nonstandard` results in an error.
+    # Set the step size by minimising an upper bound on the error of the estimate.
+    C₁ = ε * sum(abs, m.coefs)
+    C₂ = M * sum(n -> abs(m.coefs[n] * m.grid[n]^p), eachindex(m.coefs)) / factorial(p)
+    h = convert(T, min((q / (p - q) * C₁ / C₂)^(1 / p), max_step))
```

Comment:
This change gives the same performance improvement as https://github.com/JuliaDiff/FiniteDifferences.jl/pull/109#discussion_r497451612
but doesn't say anything about the `accuracy` argument
```suggestion
    # type-inference fails on this, so we annotate it, which gives big performance benifits
    h::T = convert(T, min((q / (p - q) * C₁ / C₂)^(1 / p), max_step))
```